File: solutions/1382. Balance a Binary Search Tree/1382.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def balanceBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
    nums = []

    def inorder(root: Optional[TreeNode]) -> None:
      if not root:
        return
      inorder(root.left)
      nums.append(root.val)
      inorder(root.right)

    inorder(root)

    def build(l: int, r: int) -> Optional[TreeNode]:
      if l > r:
        return None
      m = l + (r - l) // 2
      return TreeNode(nums[m], build(l, m - 1), build(m + 1, r))

    return build(0, len(nums) - 1)

File: solutions/1382. Balance a Binary Search Tree/1382.rs
// Time complexity: O(n)
// Space complexity: O(n)
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn balance_bst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut nums = Vec::new();
        Self::inorder(&root, &mut nums);
        Self::build(&nums, 0, nums.len() as i32 - 1)
    }

    fn inorder(root: &Option<Rc<RefCell<TreeNode>>>, nums: &mut Vec<i32>) {
        if let Some(node) = root {
            let node = node.borrow();
            Self::inorder(&node.left, nums);
            nums.push(node.val);
            Self::inorder(&node.right, nums);
        }
    }

    fn build(nums: &Vec<i32>, l: i32, r: i32) -> Option<Rc<RefCell<TreeNode>>> {
        if l > r {
            return None;
        }
        let m = (l + r) / 2;
        let node = Rc::new(RefCell::new(TreeNode::new(nums[m as usize])));
        node.borrow_mut().left = Self::build(nums, l, m - 1);
        node.borrow_mut().right = Self::build(nums, m + 1, r);
        Some(node)
    }
}
